### DISCUSSION TITLE
Good way to create repository with non-OSGi source jars

### DISCUSSION THREAD

**Discussion by ptziegler** [eclipse-tycho/tycho#4052] [Q&A] [answered]
I'm currently trying to a build repository containing all 3rd party plugins (as well as their sources) that we use for our application, using the `tycho-source-plugin`. I have defined a feature containing all those plugins and a repository, only containing the generated source feature. For the majority of bundles, this already works. But some entries seem to be missing their source jars, even though they are available in the IDE.

![image](https://github.com/eclipse-tycho/tycho/assets/70652978/c5853968-dc69-46da-9db5-eaf34c53deaa)

My current guess is that all sources are omitted, that lack certain OSGi headers (most likely `Eclipse-SourceBundle:`). But even then, I also noticed that for some Orbit bundles, the source bundles are also missing, even though they are available on the update site with what I assume to be valid headers (e.g. com.sun.jna from [here](https://download.eclipse.org/justj/?file=tools/orbit/simrel/orbit-aggregation/release/4.32.0/plugins)).

Is there a general way to create those missing elements? For the Maven artifacts, it should in principle be possible to check, whether a source jar is available. So I would assume that only "only" needs to rebundle the sources to make this work.

**Comment by laeubi** [comment, marked as answer]
In general there is no need for any special "building":

1. create a target platform
2. add all dependencies as maven locations (**and enable to include sources!**)
3. create a `category.xml` and add all bundles (without sources!)
4. enable [includeAllSources](https://tycho.eclipseprojects.io/doc/latest/tycho-p2-repository-plugin/assemble-repository-mojo.html#includeAllSources)

If that don't work, please explain why and we might find a solution :-)

After that, think about if you really want such a site, there are usually better alternatives today (e.g. simply include them in your repository for the product using repository filtering) or just not mirror the item in the first place, e.g. PDE + m2e can already gather sources on demand.

**Reply by merks** [reply]
Orbit is doing that here:

https://github.com/eclipse-orbit/orbit-simrel/tree/main/maven-osgi

It's handy to group things in a feature so you don't have to list all the bundles in the category.xml.

**Reply by ptziegler** [reply]
> create a target platform
> add all dependencies as maven locations (and enable to include sources!)
> create a category.xml and add all bundles (without sources!)

That was already my rough setup. I reuse the same target platform that is used for the product. I already had the category set up, which only had our dependency feature as only element. Given our 380 3rd-party dependencies, it doesn't make sense to work directly on the bundles, but that shouldn't have any impact on the final result.

> enable [includeAllSources](https://tycho.eclipseprojects.io/doc/latest/tycho-p2-repository-plugin/assemble-repository-mojo.html#includeAllSources)

That seems to have done the trick and the second round looks much better. Thanks a lot! 😄

![image](https://github.com/eclipse-tycho/tycho/assets/70652978/30d1d5e2-2fb2-407c-834c-ffbb01b3eabf)

I also noticed that I had set `targetDefinitionIncludeSource` to `ignore` because we had a source jar with an invalid manifest at some point. which also might've contributed to this behavior.

So in short: I no longer create the source feature and simply let Tycho do all the work for me :)
